src/runtime/utils/form.ts

```diff
@@ -14,109 +14,198 @@ export function isYupError(error: any): error is YupError {
+  }
 }
 
-export async function getJoiErrors(state: any, schema: JoiSchema): Promise<FormError[]> {
+async function validateJoiSchema(
+  state: any,
+  schema: JoiSchema
+): Promise<ValidateReturnSchema<typeof state>> {
   try {
     await schema.validateAsync(state, { abortEarly: false })
-    return []
+    return {
+      errors: null,
+      result: state
+    }
   } catch (error) {
     if (isJoiError(error)) {
-      return error.details.map(detail => ({
-        name: detail.path.join('.'),
-        message: detail.message
+      const errors = error.details.map(issue => ({
+        name: issue.path.join('.'),
+        message: issue.message
       }))
+
+      return {
+        errors,
+        result: null
+      }
     } else {
       throw error
     }
   }
 }
 
-export function isValibotSchema(schema: any): schema is ValibotSchema | ValibotSchemaAsync | ValibotSafeParser<any, any> | ValibotSafeParserAsync<any, any> {
-  return '_run' in schema || (typeof schema === 'function' && 'schema' in schema)
-}
-
-export async function getValibotErrors(
+async function validateValibotSchema(
   state: any,
   schema: ValibotSchema | ValibotSchemaAsync | ValibotSafeParser<any, any> | ValibotSafeParserAsync<any, any>
-): Promise<FormError[]> {
+): Promise<ValidateReturnSchema<typeof state>> {
   const result = await ('_run' in schema ? schema._run({ typed: false, value: state }, {}) : schema(state))
-  return result.issues?.map(issue => ({
-    // We know that the key for a form schema is always a string or a number
-    name: issue.path?.map((item: any) => item.key).join('.') || '',
-    message: issue.message
-  })) || []
-}
 
-export function isStandardSchema(schema: any): schema is StandardSchema {
-  return '~standard' in schema
-}
+  if (!result.issues || result.issues.length === 0) {
+    const output = ('output' in result
+      ? result.output
+      : 'value' in result
+        ? result.value
+        : null)
+    return {
+      errors: null,
+      result: output
+    }
+  }
 
-export async function getStandardErrors(
-  state: any,
-  schema: StandardSchema
-): Promise<FormError[]> {
-  const result = await schema['~validate']({ value: state })
-  return result.issues?.map(issue => ({
-    name: issue.path?.map(item => typeof item === 'object' ? item.key : item).join('.') || '',
+  const errors = result.issues.map(issue => ({
+    name: issue.path?.map((item: any) => item.key).join('.') || '',
     message: issue.message
-  })) || []
+  }))
+
+  return {
+    errors,
+    result: null
+  }
 }
 
-export async function getSuperStructErrors(state: any, schema: Struct<any, any>): Promise<FormError[]> {
-  const [err] = schema.validate(state)
-  if (err) {
-    const errors = err.failures()
-    return errors.map(error => ({
-      message: error.message,
-      name: error.path.join('.')
-    }))
+export function parseSchema<T extends object>(state: T, schema: FormSchema<T>): Promise<ValidateReturnSchema<typeof state>> {
```

Comment:
Could you add some tests for the transform logic?

Reply:
Okay, I forget since v3 it should be implement `tests` @romhml

Reply:
No problem! Thanks!

Reply:
i already make test form each schema validation. But Unfortunately, `superstruct` doesn't provide a `transform` method after validation, so I decided not to include it in the Form test.